Control enhancements

kchobantonov opened the pull request:
fixes #90 

Allow allOf to also show properties if any, allow the object to display additionalProperties, patternProperties and conform with the propertyNames schema when defining the property names. The Add and Delete buttons will be disable when number of properties is greater or lower than what is specified in maxProperties/minProperties

[Additional Properies](https://json-schema.org/understanding-json-schema/reference/object.html#additional-properties)
[Pattern Properties](https://json-schema.org/understanding-json-schema/reference/object.html#pattern-properties)
[Property Names](https://json-schema.org/understanding-json-schema/reference/object.html#property-names)
[Size](https://json-schema.org/understanding-json-schema/reference/object.html#size)

kchobantonov commented:
@sdirix will you have time soon to review this change and have a new release ?

kchobantonov commented:
@sdirix  @lucas-koehler does any of you have the time this week to review and merge this PR ?

sdirix commented:
Hi @kchobantonov,

@lucas-koehler will review this PR this week. Depending on the number of required iterations this might then get in this week (or not). Once it's merged I can do a prerelease.

We love working on JSON Forms, but we only have limited time each month to handle topics which are not related to any of our customer projects. You are a very active community contributor, however the contributions are mostly limited to the Vuetify renderer set [which is really only a minor part of the overall JSON Forms offerings](https://npmtrends.com/@jsonforms/angular-material-vs-@jsonforms/core-vs-@jsonforms/material-renderers-vs-@jsonforms/vanilla-renderers-vs-@jsonforms/vue-vanilla-vs-@jsonforms/vue2-vanilla-vs-@jsonforms/vue2-vuetify). We therefore can't give full priority to all of your incoming contributions (at the very moment 4! large PRs simultaneously). Of course we will still look at them, but one at a time.

For the future we will try to communicate more clearly when and what we are looking next at. Once a reviewer is assigned to one of your PRs we look at it very soon. On your side you can try to polish the PRs as much as possible to help us out before requesting a review. You can use draft PRs if needed to mark that something is not yet ready to be reviewed. The smaller a PR is, the easier it's also for us to schedule a review. The less iterations and the less time reviewing your contributions take, the more we can potentially take on.

lucas-koehler commented on vue2-vuetify/src/complex/OneOfTabRenderer.vue:
```diff
@@ -0,0 +1,188 @@
+  },
+  setup(props: RendererProps<ControlElement>) {
+    const input = useJsonFormsOneOfControl(props);
+    const control = (input.control as any).value as typeof input.control;
```
Why can you not use `input.control` directly?

kchobantonov replied:
the input.control is actually a a wrapper and inside the javascript/typescript those wrappers are not automatically outboxed - e.g. in the setup for ref objects you do need to used v.value instead just v for example.

lucas-koehler commented on vue2-vuetify/src/controls/DateControlRenderer.vue:
```diff
@@ -155,9 +158,19 @@ const controlRenderer = defineComponent({
-      return typeof this.appliedOptions.pickerIcon == 'string'
-        ? this.appliedOptions.pickerIcon
-        : 'mdi-calendar';
+      if (typeof this.appliedOptions.pickerIcon == 'string') {
```
Please use strict comparisons to avoid unexpected bugs :)
```suggestion
      if (typeof this.appliedOptions.pickerIcon === 'string') {
```

lucas-koehler commented on vue2-vuetify/src/controls/DateControlRenderer.vue:
```diff
@@ -155,9 +158,19 @@ const controlRenderer = defineComponent({
+        return this.appliedOptions.pickerIcon;
+      }
+
+      if (this.pickerType == 'year') {
```
Please use strict comparisons to avoid unexpected bugs :)
```suggestion
      if (this.pickerType === 'year') {
```

lucas-koehler commented on vue2-vuetify/src/controls/DateControlRenderer.vue:
```diff
@@ -155,9 +158,19 @@ const controlRenderer = defineComponent({
+        return 'mdi-alpha-y-box-outline';
+      }
+
+      if (this.pickerType == 'month') {
```
Please use strict comparisons to avoid unexpected bugs :)
```suggestion
      if (this.pickerType === 'month') {
```

lucas-koehler commented on vue2-vuetify/src/complex/components/AdditionalProperties.vue:
```diff
@@ -0,0 +1,463 @@
+              :aria-label="addToLabel"
+              v-on="onTooltip"
+              :disabled="
+                !control.enabled ||
+                (appliedOptions.restrict &&
+                  control.schema.maxProperties !== undefined &&
+                  control.data &&
+                  Object.keys(control.data).length >=
+                    control.schema.maxProperties) ||
+                errors.length > 0 ||
+                !newPropertyName
```
This computation is rather complicated and hard to read. Could you move it to code instead of instead of embedding it in the template and maybe add a short comment?

lucas-koehler replied:
@kchobantonov Please have a look again at this comment. I think it was hidden earlier.

kchobantonov replied:
added in the last commit

lucas-koehler commented on vue2-vuetify/src/complex/components/AdditionalProperties.vue:
```diff
@@ -0,0 +1,463 @@
+                small
+                :aria-label="deleteLabel"
+                :disabled="
+                  !control.enabled ||
+                  (appliedOptions.restrict &&
+                    control.schema.minProperties !== undefined &&
+                    control.data &&
+                    Object.keys(control.data).length <=
+                      control.schema.minProperties)
```
This computation is rather complicated and hard to read. Could you move it to code instead of instead of embedding it in the template and maybe add a short comment?

lucas-koehler replied:
@kchobantonov Please have a look again at this comment. I think it was hidden earlier.

kchobantonov replied:
added in the last commit

lucas-koehler commented on vue2-vuetify/src/complex/components/AdditionalProperties.vue:
```diff
@@ -0,0 +1,463 @@
+    },
+  },
+  setup(props) {
+    const control = props.input.control as any as Ref<
```
Why do you need to cast this here? Couldn't you instead adapt the type of the prop instead?

lucas-koehler replied:
@kchobantonov Please have a look again at this comment. I think it was hidden earlier.

kchobantonov replied:
because the control is a reference and as such need to be accessed with .value - the IDE just doesn't know that the actual value that is going to be passed is not of type T but Ref<T> and if I do not do that type cast then the rest of the code the IDE will not give me an error if I access the properties directly instead via the .value property as it should when I'm in the setup function - the Vue does not automatically do the property access for ref objects in the setup like that is happening in the methods, computed values or the template.

lucas-koehler commented on vue2-vuetify/src/complex/components/AdditionalProperties.vue:
```diff
@@ -0,0 +1,463 @@
+    const additionalPropertyItems = ref<AdditionalPropertyType[]>([]);
+
+    additionalKeys.forEach((propName) => {
+      const additionaProperty = toAdditionalPropertyType(
```
Fix typo
```suggestion
      const additionaProperty = toAdditionalPropertyType(
```

lucas-koehler commented on vue2-vuetify/src/complex/components/AdditionalProperties.vue:
```diff
@@ -0,0 +1,463 @@
+      typeof control.value.schema.additionalProperties !== 'object' &&
+      typeof control.value.schema.patternProperties === 'object'
+    ) {
+      const matchPatternPropeertiesKeys: JsonSchema7 = {
```
Fix typo ;)
```suggestion
      const matchPatternPropertiesKeys: JsonSchema7 = {
```

lucas-koehler commented on vue2-vuetify/src/complex/components/AdditionalProperties.vue:
```diff
@@ -0,0 +1,463 @@
+    const vuetifyProps = (path: string) => {
+      const props = get(appliedOptions.value?.vuetify, path);
+
+      return props && isPlainObject(props) ? props : {};
```
Why is the `isPlainObject` check needed?

kchobantonov replied:
this matched the vuetifyProps that we have currently in the composition file - there we also check the isPlainObject which I guess could be removed but it was checking that to guarantee that this is an object and not a simple data like String or number for example - e.g. with incorrect JSON files.

lucas-koehler commented on vue2-vuetify/src/complex/components/AdditionalProperties.vue:
```diff
@@ -0,0 +1,463 @@
+              ap.schema
+            ) {
+              hasChanges = true;
+              newData[ap.propertyName] = createDefaultValue(ap.schema);
```
What happens if the default value is `undefined` or `null`? Could this cause an endless loop?

kchobantonov replied:
yes you are right - in the special case where the schema type is 'null' then this can happen - will add code fix for that

kchobantonov replied:
resolved in the latest commit

kchobantonov replied:
sure thing

lucas-koehler commented on example/src/store/validate/dynamicDefaults.ts:
```diff
@@ -0,0 +1,149 @@
+import dayjs, { UnitType } from 'dayjs';
+import customParsingPlugin from 'dayjs/plugin/customParseFormat';
+import durationPlugin from 'dayjs/plugin/duration';
+import timezonePlugin from 'dayjs/plugin/timezone'; // dependent on utc plugin
+import utcPlugin from 'dayjs/plugin/utc';
+
+// required for the custom save formats in the date, time and date-time pickers
+dayjs.extend(customParsingPlugin);
+dayjs.extend(utcPlugin);
+dayjs.extend(timezonePlugin);
+dayjs.extend(durationPlugin);
```
Would other applications using this renderer set also need to configure this? If yes, it should be documented in the README

lucas-koehler reviewed (commented):
Hey @kchobantonov ,
I had a look at the code. See inline comments.
Could you briefly explain why we need to extract the css because of the vue shadow dom? Also, I assume users would need to add this extracted css file to their index.html (or load it some other way), right?

I did not test the changes so far. I will do this later this week.

kchobantonov commented on vue2-vuetify/src/complex/OneOfRenderer.vue:
```diff
@@ -111,22 +118,26 @@ const controlRenderer = defineComponent({
 
-    const selectedIndex = ref(control.indexOfFittingSchema || 0);
-    const tabIndex = ref(selectedIndex.value);
+    const selectedIndex = ref(control.indexOfFittingSchema);
```
because is always selects the first one and in case where we have a dropdown now and the property is required we need explicitly to be selected by the end user - otherwise the dropdown will be automatically selected to the first one

kchobantonov commented:
@lucas-koehler thanks for the review - all suggestion that are now in the code as fixes I have marked those as resolve conversation - the only left here are those with the comments and why we need those. let me know if I can clarify some of those more.

kchobantonov commented:
> Hey @kchobantonov , I had a look at the code. See inline comments. Could you briefly explain why we need to extract the css because of the vue shadow dom? Also, I assume users would need to add this extracted css file to their index.html (or load it some other way), right?
> 
> I did not test the changes so far. I will do this later this week.

Yes they need to add the css the same way how the App.vue in the example application is adding that - e.g. 

```
@import '~@jsonforms/vue2-vuetify/lib/jsonforms-vue2-vuetify.esm.css';
```

And this is needed because otherwise the inclusion of the CSS happens in the javascript code as it is using a default Vue CSS injector which injects the CSS on the document level which does not work when this is used as webcomponents. So it is better to be explicit what needs to be included where. Also all other major libraries have their CSS split so you can not only see how the CSS looks like but also could be used for CDN and etc.

lucas-koehler reviewed (changes requested):
@kchobantonov Thanks you for the updates and explanations. I resolved the corresponding conversations.
Please have a look at the unresolved ones.

I'm fine with the CSS extraction but please add information about this to the README at `vue2-vuetify/README.me`, e.g. in the code example.
@sdirix Any objections to the CSS being extracted and bundled into a separate CSS file?

kchobantonov commented:
> @kchobantonov Thanks you for the updates and explanations. I resolved the corresponding conversations. Please have a look at the unresolved ones (some of them were in the hidden ones when I opened the PR again).
> 
> I'm fine with the CSS extraction but please add information about this to the README at `vue2-vuetify/README.me`, e.g. in the code example. @sdirix Any objections to the CSS being extracted and bundled into a separate CSS file?



> @kchobantonov Thanks you for the updates and explanations. I resolved the corresponding conversations. Please have a look at the unresolved ones (some of them were in the hidden ones when I opened the PR again).
> 
> I'm fine with the CSS extraction but please add information about this to the README at `vue2-vuetify/README.me`, e.g. in the code example. @sdirix Any objections to the CSS being extracted and bundled into a separate CSS file?

@lucas-koehler all suggested changes are included in the last commit

kchobantonov commented:
@sdirix  @lucas-koehler  the only outstanding issue that I can see is with the Number and Integer controls where if for example for Number control [control](https://deploy-preview-91--jsonforms-vuetify-renderers.netlify.app/#/example/control) put a value for example `50.1234567890123441231313131313132` then in the actual JSON data we will have the number `50.123456789012344` which is correct but then the UI and the JSON data they do not exactly match. We have I guess 2 options  here - either return the string if the value does not match with the String but that would be tricky since the parse number might be the same as number as the input string but written in exponent syntax (with E) or the second option is to synch the input with whatever we have successfully parsed as float - e.g. if parseFloat is truthy then to schedule an update for the UI so that the UI represent exactly what we have as data otherwise it looks strange... Not that if you paste the first number in the Number input the first time the number will be truncated to the second value but then you can still enter more numbers in the input

So which approach we should take ?

lucas-koehler commented:
@kchobantonov Thanks for the update. There were actually 3 more comments from the first review. Sorry if I was not clear about that. I commented them again and tagged you so that you can find them.

> the only outstanding issue that I can see is with the Number and Integer controls where if for example for Number [control](https://deploy-preview-91--jsonforms-vuetify-renderers.netlify.app/#/example/control) we put a value for example 50.1234567890123441231313131313132 then in the actual JSON data we will have the number 50.123456789012344 which is correct but then the UI and the JSON data they do not exactly match. We have I guess 2 options here - either return the string if the value does not match with the String but that would be tricky since the parse number might be the same as number as the input string but written in exponent syntax (with E) or the second option is to synch the input with whatever we have successfully parsed as float - e.g. if parseFloat is truthy then to schedule an update for the UI so that the UI represent exactly what we have as data otherwise it looks strange... Note that if you paste the first number in the Number input the first time the number will be truncated to the second value but then you can still enter more numbers in the input

I think we should go with the second approach: The UI should resemble the data as close as possible (aside from local specific formatting).

kchobantonov commented:
@lucas-koehler I have removed for Number and Integer controls the type=number because it was working strange under chrome - for example when you enter - `E1211` - or - `1.7976931348623157e+309` - which is larger than the max number `1.7976931348623157e+308` the string returned by the input was empty string - if I remove the type=number then it returns the actual string that is used. 

Also I went with the case where if we can't convert to number then we pass the string and let the AJV handle the error that this is not a number.  Since this is not input with type=number the control allows entering any characters so that is why we used not only parseFloat but also Number() parsing which is more strict but fails to return NaN for empty or strings with only whitespaces - that is why we use the combination of the two.

lucas-koehler commented:
@kchobantonov Thanks for the updates again :)
Your change for the numbers sound reasonable to me. I tested that an found one issue:
If you enter a number that is too long.
E.g. `5675.4444444444444444444444444444444`
This then gets shortened to: `5675.444444444444` which is as we want it.
When I now add another digit to the input field (e.g. a `5`) the input shows `5675.4444444444445` but the data stays at `5675.444444444444`.
I guess what happens is that the UI doesn't get updated again because the data value stays the same.
Could you have a look at that again?

kchobantonov commented:
@lucas-koehler  I have updated the integer and number controls to handle correctly 0 - there was again this issue with the truthy values like with the boolean controls. I have also make sure that we have a dedicated local data variable inputValue to hold what the user entered and the reason is that if we directly use the converted number then this number might get converted to a exponent form when the value is large and also the second reason is that we have removed the type=number by default (again the user can specify that in the UI schema as additional vuetify attributes if desired) 

The integer control also have a new option - `allowUnsafeInteger` - that is boolean and by default is false which means that the integer number that we get needs to pass the check Number.isSafeInteger before converting it to a number. The reason for that is that when we type integers out of that range then the number that we type on the input can be different than the number stored in the data. The end users could opt out from that behavior if he/she specify `allowUnsafeInteger` to true to allow such integers.

I have removed the Number() call to validate additionally to the parseFloat if the value is valid and instead using regex to validate the input before converting it to number using parseFloat because of how parseFloat is implemented.

The issue that you have outlined above with `5675.4444444444444444444444444444444` should be fixes as well

lucas-koehler reviewed (approved):
LGTM. Thanks again for the contribution, the updates and explanation :)
